### PR TITLE
SoftmaxCrossEntropyLossInternalGrad and Sum Fusion

### PR DESCRIPTION
**Pull request by iK1D**
We observed the SoftmaxCrossEntropyLossInternalGrad+Sum pattern in more than one customer models, both of the nodes need ~10ms to compute when the input tensor shape is relatively big, especially with big vocab size. The PR is to fuse this two Ops to a single one. In CUDA/ROCm EP, only one fused Ops is launched, so that the total execution is reduced by half. We can also observe >2% perf gain for the whole model from the throughput.

### CODE REVIEW

File: orttraining/orttraining/core/optimizer/sce_loss_grad_bias_fusion.cc [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    Node& sum_node = *p_next;
+    auto& sum_input_defs = sum_node.MutableInputDefs();
+    auto shape0 = sum_input_defs[0]->Shape();
+    auto shape1 = sum_input_defs[1]->Shape();
+    if (!shape0 || !shape1 || shape0->dim_size() != shape1->dim_size()) {
+      continue;
+    }
+
+    bool has_same_shape = true;
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/sce_loss_grad_bias_fusion.cc:59]: I recalled Shape can be compared with ==. 

```
>   if (input->Shape() != skip->Shape()) {
>     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
>                            "skip is expected to have same shape as input");
>   }
```

File: orttraining/orttraining/core/optimizer/sce_loss_grad_bias_fusion.cc [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+
+namespace onnxruntime {
+
+Status SceLossGradBiasFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/sce_loss_grad_bias_fusion.cc:15]: the fusion graph is looks like this? 

SoftmaxCrossEntropyLossInternalGrad --> optional Reshape --> Add|Sum ?

File: orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc [hunk 3]
```diff
@@ -386,6 +386,16 @@ Status SoftmaxCrossEntropyLossGrad<T1, T2>::Compute(OpKernelContext* context) co
     d_logit->Reshape(new_shape);
   }
 
+  // Bias.
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc:389]: I am a bit surprised we did not use parallel for here. maybe for the newly added elementwise add, it's simple to a parallel loop.

File: orttraining/orttraining/test/optimizer/graph_transform_test.cc [hunk 4]
```diff
@@ -199,6 +201,222 @@ TEST_F(GraphTransformationTests, DropoutWithZeroRatioElimination) {
...
+        auto& attrs = node.GetAttributes();
+        ASSERT_TRUE(attrs.find("reduction") != attrs.end());
+        ASSERT_EQ(reduction, attrs.at("reduction").s());
+        ASSERT_EQ(6, static_cast<int>(node.InputDefs().size()));
+      }
+    }
+  };
+
+  std::unique_ptr<GraphTransformer> transformer = std::make_unique<SceLossGradBiasFusion>();
+  TestGraphTransformer(build_test_case, 14, logger, std::move(transformer), TransformerLevel::Level2, 1,
```
Comment (pengwa) [inline orttraining/orttraining/test/optimizer/graph_transform_test.cc:278]: shall we also run opset 12, 13, we had models onboarded running on 12, they probably will get refreshed to re-train with new ORT + new data, benefiting from this fuse.

**Comment by pengwa** [comment]
FYI @baijumeswani . let's check whether on-device training models (GPU/CPU) have this pattern or not.

File: orttraining/orttraining/core/optimizer/sce_loss_grad_bias_fusion.cc (hunk 1 above)
Comment (iK1D) [reply orttraining/orttraining/core/optimizer/sce_loss_grad_bias_fusion.cc:59]: There are different "shape". ORT's tensor shape is onnxruntime::TensorShape, the "shape" in ONNX graph (in transformer code) is onnx::TensorShapeProto.

File: orttraining/orttraining/core/optimizer/sce_loss_grad_bias_fusion.cc (hunk 2 above)
Comment (iK1D) [reply orttraining/orttraining/core/optimizer/sce_loss_grad_bias_fusion.cc:15]: Right

File: orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc (hunk 3 above)
Comment (iK1D) [reply orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc:389]: It may means no body use CPU for training, so I didn't take effort to refactor the code, and just to make it work.

Comment (iK1D) [reply orttraining/orttraining/training_ops/cpu/loss/softmax_cross_entropy_loss.cc:389]: If it's critical for on-device training, we can optimize this in a new PR.

File: orttraining/orttraining/test/optimizer/graph_transform_test.cc (hunk 4 above)
Comment (iK1D) [reply orttraining/orttraining/test/optimizer/graph_transform_test.cc]: Added the test. But the old ORT will not have this fusion. If user wants to re-train with new ORT with this fusion, the default OpSet version for ORTModule is now 15, unless user uses env variables to set the OpSet version to an old one, which is not recommended.

**Review by zhijxu-MS — APPROVED** [review]
